crates/core/src/subscription/subscription.rs

```diff
@@ -744,6 +501,280 @@ fn with_delta_table(mut join: IndexJoin, index_side: bool, delta: DatabaseTableU
+impl ExecutionUnit<SupportedQuery> {
+    fn eval(&self, db: &RelationalDB, tx: &Tx, auth: AuthCtx) -> Result<DatabaseTableUpdate, DBError> {
+        let ctx = ExecutionContext::subscribe(db.address());
+        let mut ops = Vec::new();
+
+        for table in run_query(&ctx, db, tx, &self.2.expr, auth)? {
+            for row in table.data {
+                let op_type = 1;
+                let row_pk = pk_for_row(&row).to_bytes();
+                let row = row.data;
+                ops.push(TableOp { op_type, row_pk, row });
+            }
+        }
+
+        Ok(DatabaseTableUpdate {
+            table_id: self.0,
+            table_name: self.1.to_owned(),
+            ops,
+        })
+    }
+
+    fn eval_incr(
+        &self,
+        db: &RelationalDB,
+        tx: &Tx,
+        database_update: &DatabaseUpdate,
+        auth: AuthCtx,
+    ) -> Result<DatabaseTableUpdate, DBError> {
+        let table_id = self.0;
+        let table_name = self.1.to_owned();
+
+        let kind = self.2.kind;
+        let expr = self.2.as_expr();
+
+        use query::Supported::*;
+        let ops = match kind {
+            Scan => {
+                if let Some(rows) = database_update.tables.iter().find(|update| update.table_id == self.0) {
+                    // Replace table reference in original query plan with virtual MemTable
+                    let plan = query::to_mem_table(expr.clone(), rows);
+                    let eval = evaluator_for_primary_updates(db, auth);
+                    // Evaluate the new plan and capture the new row operations
+                    eval(tx, &plan)?.into_iter().map(|op| op.into()).collect()
+                } else {
+                    vec![]
+                }
+            }
+            Semijoin => {
+                if let Some(plan) = IncrementalJoin::new(expr, database_update.tables.iter())? {
+                    // Evaluate the plan and capture the new row operations
+                    plan.eval(db, tx, &auth)?.map(|op| op.into()).collect()
+                } else {
+                    vec![]
+                }
+            }
+        };
+        Ok(DatabaseTableUpdate {
+            table_id,
+            table_name,
+            ops,
+        })
+    }
+}
+
+/// An ExecutionUnit can be a set of queries that return rows from the same table.
+/// In this case we need to deduplicate the result set.
+impl ExecutionUnit<BTreeSet<SupportedQuery>> {
+    fn eval(&self, db: &RelationalDB, tx: &Tx, auth: AuthCtx) -> Result<DatabaseTableUpdate, DBError> {
+        let ctx = ExecutionContext::subscribe(db.address());
+        let mut ops = Vec::new();
+        let mut dup = HashSet::new();
+
+        for SupportedQuery { kind: _, expr } in &self.2 {
+            for table in run_query(&ctx, db, tx, expr, auth)? {
+                for row in table.data {
+                    let row_pk = pk_for_row(&row);
+
+                    // Skip duplicate rows
+                    if dup.contains(&row_pk) {
```

Comment:
Wait, is possible to get the same `row_pk` by different tables? because the comment say it need to dedup by table, not by row.

Reply:
It is possible, but that's fine, because the same row in two different tables is considered two different rows.